tests/test_class.cpp

```diff
@@ -231,7 +231,8 @@ TEST_SUBMODULE(class_, m) {
         };
 
         auto def = new PyMethodDef{"f", f, METH_VARARGS, nullptr};
-        return py::reinterpret_steal<py::object>(PyCFunction_NewEx(def, nullptr, m.ptr()));
+        py::capsule def_capsule(def, [](void *ptr) { delete reinterpret_cast<PyMethodDef *>(ptr); });
+        return py::reinterpret_steal<py::object>(PyCFunction_NewEx(def, def_capsule.ptr(), m.ptr()));
```

Comment:
`PyCFunction_NewEx()` just calls [`PyCMethod_new()`](https://github.com/python/cpython/blob/04b8631d84a870dda456ef86039c1baf34d08500/Objects/methodobject.c#L44) with the last parameter set to NULL.

What this changes is [`m_self`](https://github.com/python/cpython/blob/04b8631d84a870dda456ef86039c1baf34d08500/Objects/methodobject.c#L110) from NULL to `def_capsule.ptr()`.

___

Beyond that, I have no idea how this works. `PyCFunction*` aren't documented, which is discussed in [bpo-16776](https://github.com/python/cpython/blob/04b8631d84a870dda456ef86039c1baf34d08500/Objects/methodobject.c#L110).

Reply:
> What this changes is [`m_self`](https://github.com/python/cpython/blob/04b8631d84a870dda456ef86039c1baf34d08500/Objects/methodobject.c#L110) from NULL to `def_capsule.ptr()`.

We already do this in the actual pybind11 code, btw.

Alternatively, if you prefer, we could play the "`weakref` trick", here? But I thought this would be easier, since AFAICT `m_self` being `NULL` wasn't the point of this test.

Reply:
I'm just trying to understand how does setting `m_self` work to our benefit in this case. Does it steal the reference or borrow it? Is our reference counting right in this case? If it appeased valgrind, I guess it's correct, I'd just like a better understanding of how it works.

Reply:
It [does borrow](https://github.com/python/cpython/blob/04b8631d84a870dda456ef86039c1baf34d08500/Objects/methodobject.c#L108-L109), yes, and keeps it alive as some kind of CPython-style `void *data`, passed as first argument to our `cpp_function::dispatcher`: https://github.com/pybind/pybind11/blob/d587a2fd17dc89c0b2d4b6dc7715594bed45fb58/include/pybind11/pybind11.h#L492

See here for the other case where CPython uses it:
https://github.com/pybind/pybind11/blob/d587a2fd17dc89c0b2d4b6dc7715594bed45fb58/include/pybind11/pybind11.h#L372

This mechanism was also our problem with Python 3.9.0, `m_self` being decref'ed a few lines too early while pybind11 was still accessing it, in #2576.